File: sdks/java/src/main/java/com/axora/logger/WideEvent.java
package com.axora.logger;

import com.axora.logger.sinks.Sink;
import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import java.io.PrintWriter;
import java.io.StringWriter;
import java.time.Instant;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.UUID;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionException;

public class WideEvent {
  private static final String SDK_VERSION = "0.1.0";
  private static final ObjectMapper MAPPER = new ObjectMapper();

  private final String eventId = UUID.randomUUID().toString();
  private final Instant timestampStart = Instant.now();
  private final long monotonicStartNanos = System.nanoTime();
  private final String operation;
  private final String service;
  private final String environment;
  private final List<Sink> sinks;
  private final Map<String, Object> context = new LinkedHashMap<>();
  private WideEventPayload.ErrorInfo error = new WideEventPayload.ErrorInfo(null, null, null);
  private String level = "info";
  private String status = "ok";
  private boolean finalized = false;
  private CompletableFuture<Void> emitFuture;

  WideEvent(String operation, String service, String environment, List<Sink> sinks) {
    this.operation = operation;
    this.service = service;
    this.environment = environment;
    this.sinks = List.copyOf(sinks);
  }

  private void ensureMutable() {
    if (finalized) {
      throw new IllegalStateException("WideEvent has already been finalized");
    }
  }

  public WideEvent appendContext(Map<String, Object> fields) {
    ensureMutable();
    context.putAll(deepCopy(fields));
    return this;
  }

  public WideEvent setError(Throwable throwable) {
    ensureMutable();
    level = "error";
    status = "error";
    error =
        new WideEventPayload.ErrorInfo(
            throwable.getClass().getSimpleName(),
            throwable.getMessage(),
            stackTraceToString(throwable));
    return this;
  }

  private Map<String, Object> deepCopy(Map<String, Object> fields) {
    return MAPPER.convertValue(fields, new TypeReference<LinkedHashMap<String, Object>>() {});
  }

  private static String stackTraceToString(Throwable throwable) {
    StringWriter sw = new StringWriter();
    throwable.printStackTrace(new PrintWriter(sw));
    return sw.toString();
  }

  private WideEventPayload buildPayload(String overrideLevel, String overrideStatus) {
    Map<String, Object> snapshotContext = deepCopy(context);
    WideEventPayload payload =
        new WideEventPayload(
            eventId,
            service,
            environment,
            timestampStart.toString(),
            Instant.now().toString(),
            Math.round(((System.nanoTime() - monotonicStartNanos) / 1_000_000.0) * 1000.0) / 1000.0,
            overrideLevel != null ? overrideLevel : level,
            operation,
            overrideStatus != null ? overrideStatus : status,
            snapshotContext,
            error,
            new WideEventPayload.MetaInfo(SDK_VERSION, "java"));
    return payload;
  }

  private CompletableFuture<Void> safeExport(Sink sink, WideEventPayload payload) {
    try {
      return sink.export(payload).exceptionally(ex -> null);
    } catch (Throwable throwable) {
      return CompletableFuture.completedFuture(null);
    }
  }

  public CompletableFuture<Void> emitAsync() {
    return emitAsync(null, null);
  }

  public CompletableFuture<Void> emitAsync(String overrideLevel, String overrideStatus) {
    if (emitFuture != null) {
      return emitFuture;
    }

    WideEventPayload payload = buildPayload(overrideLevel, overrideStatus);
    finalized = true;
    CompletableFuture<?>[] tasks =
        sinks.stream().map(sink -> safeExport(sink, payload)).toArray(CompletableFuture[]::new);
    emitFuture = CompletableFuture.allOf(tasks);
    return emitFuture;
  }
}

File: sdks/csharp/Openakta.Logger/WideEvent.cs
using System.Collections.Concurrent;
using System.Diagnostics;
using System.Text.Json;
using Openakta.Logger.Sinks;

namespace Openakta.Logger;

public sealed class WideEvent
{
    private const string SdkVersion = "0.1.0";

    private readonly string _eventId = Guid.NewGuid().ToString();
    private readonly DateTimeOffset _timestampStart = DateTimeOffset.UtcNow;
    private readonly Stopwatch _stopwatch = Stopwatch.StartNew();
    private readonly string _operation;
    private readonly string _service;
    private readonly string _environment;
    private readonly IReadOnlyList<ISink> _sinks;
    private readonly ConcurrentDictionary<string, object?> _context = new();
    private WideEventPayload.ErrorDetail _error = new(null, null, null);
    private string _level = "info";
    private string _status = "ok";
    private bool _finalized;
    private Task? _emitTask;

    internal WideEvent(string operation, string service, string environment, IReadOnlyList<ISink> sinks)
    {
        _operation = operation;
        _service = service;
        _environment = environment;
        _sinks = sinks;
    }

    private void EnsureMutable()
    {
        if (_finalized)
        {
            throw new InvalidOperationException("WideEvent has already been finalized");
        }
    }

    public WideEvent AppendContext(IDictionary<string, object?> fields)
    {
        EnsureMutable();
        foreach (var (key, value) in fields)
        {
            _context[key] = CloneValue(value);
        }

        return this;
    }

    public WideEvent SetError(Exception ex)
    {
        EnsureMutable();
        _level = "error";
        _status = "error";
        _error = new WideEventPayload.ErrorDetail(ex.GetType().Name, ex.Message, ex.StackTrace);
        return this;
    }

    private static object? ConvertJsonElement(JsonElement element)
    {
        return element.ValueKind switch
        {
            JsonValueKind.Object => element.EnumerateObject().ToDictionary(property => property.Name, property => ConvertJsonElement(property.Value)),
            JsonValueKind.Array => element.EnumerateArray().Select(ConvertJsonElement).ToList(),
            JsonValueKind.String => element.GetString(),
            JsonValueKind.Number when element.TryGetInt64(out var intValue) => intValue,
            JsonValueKind.Number when element.TryGetDecimal(out var decimalValue) => decimalValue,
            JsonValueKind.Number => element.GetDouble(),
            JsonValueKind.True => true,
            JsonValueKind.False => false,
            JsonValueKind.Null => null,
            _ => element.GetRawText(),
        };
    }

    private static object? CloneValue(object? value)
    {
        if (value is null)
        {
            return null;
        }

        using var document = JsonDocument.Parse(JsonSerializer.Serialize(value));
        return ConvertJsonElement(document.RootElement);
    }

    private static Dictionary<string, object?> CloneContext(IDictionary<string, object?> fields)
    {
        var cloned = new Dictionary<string, object?>();
        foreach (var (key, value) in fields)
        {
            cloned[key] = CloneValue(value);
        }

        return cloned;
    }

    private WideEventPayload BuildPayload(string? level = null, string? status = null)
    {
        var endTime = DateTimeOffset.UtcNow;
        return new WideEventPayload
        {
            EventId = _eventId,
            Service = _service,
            Environment = _environment,
            TimestampStart = _timestampStart.ToString("O"),
            TimestampEnd = endTime.ToString("O"),
            DurationMs = Math.Round(_stopwatch.Elapsed.TotalMilliseconds, 3, MidpointRounding.AwayFromZero),
            Level = level ?? _level,
            Operation = _operation,
            Status = status ?? _status,
            Context = CloneContext(_context),
            Error = _error,
            Meta = new WideEventPayload.SdkMeta(SdkVersion, "csharp"),
        };
    }

    private Task SafeExport(ISink sink, WideEventPayload payload)
    {
        try
        {
            return sink.ExportAsync(payload).ContinueWith(_ => { }, TaskContinuationOptions.ExecuteSynchronously);
        }
        catch
        {
            return Task.CompletedTask;
        }
    }

    public Task EmitAsync(string? level = null, string? status = null)
    {
        if (_emitTask is not null)
        {
            return _emitTask;
        }

        var payload = BuildPayload(level, status);
        _finalized = true;
        _emitTask = Task.WhenAll(_sinks.Select(sink => SafeExport(sink, payload)));
        return _emitTask;
    }
}
